Project references with path aliases

**TypeScript Version:**  typescript@3.4.0-dev.20190214

**Search Terms:** TS2307 `Project reference` `alias` `paths` `Cannot find module`.

**Code**  https://github.com/DataTables/project-references-demo .

The repo is cloned from https://github.com/RyanCavanaugh/project-references-demo and slightly modified to demonstrate the issue. It appears to be related to the fact that I am storing my source files in a `src` sub-directory for each component and using a `path` alias to reference between them. Running:

```
npm install
node node_modules/typescript/bin/tsc -p animals --traceResolution
```

Produces output that includes:

```
======== Resolving module '@core/utilities' from '/home/allan/Downloads/project-references-demo-master/animals/src/dog.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
'baseUrl' option is set to '/home/allan/Downloads/project-references-demo-master/', using this value to resolve non-relative module name '@core/utilities'.
'paths' option is specified, looking for a pattern to match module name '@core/utilities'.
Module name '@core/utilities', matched pattern '@core/*'.
Trying substitution 'core/src/*', candidate module location: 'core/src/utilities'.
Loading module as file / folder, candidate module location '/home/allan/Downloads/project-references-demo-master/core/src/utilities', target file type 'TypeScript'.
File '/home/allan/Downloads/project-references-demo-master/core/src/utilities.ts' exist - use it as a name resolution result.
```

indicating that the file was correctly resolved. However, I get:

```
animals/src/dog.ts:2:32 - error TS2307: Cannot find module '@core/utilities'.

2 import { makeRandomName } from '@core/utilities'
```

from tsc. The reason I've filed this as a bug is that the two don't appear to align - it is correctly resolved, but then tsc says it can't be found.

**Expected behavior:** 

No error. The aliased module is imported.

**Actual behavior:**

The TS2307 error is given. VS Code also shows the same error.

**Related Issues:**

#26823 Seems very similar, but there the module resolver does not resolve. Here it does, but it still ultimately showing as cannot be found.
#26819 _might_ be related, but I'm not sure and the suggested resolution there doesn't work.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -2352,19 +2352,12 @@ namespace ts {
             }
 
             if (moduleNotFoundError) {
-                // For relative paths, see if this was possibly a projectReference redirect
-                if (pathIsRelative(moduleReference)) {
-                    const sourceFile = getSourceFileOfNode(location);
-                    const redirects = sourceFile.redirectedReferences;
-                    if (redirects) {
-                        const normalizedTargetPath = getNormalizedAbsolutePath(moduleReference, getDirectoryPath(sourceFile.fileName));
-                        for (const ext of [Extension.Ts, Extension.Tsx]) {
-                            const probePath = normalizedTargetPath + ext;
-                            if (redirects.indexOf(probePath) >= 0) {
-                                error(errorNode, Diagnostics.Output_file_0_has_not_been_built_from_source_file_1, moduleReference, probePath);
-                                return undefined;
-                            }
-                        }
+                // See if this was possibly a projectReference redirect
+                if (resolvedModule) {
+                    const redirect = host.getProjectReferenceRedirect(resolvedModule.resolvedFileName);
+                    if (redirect) {
+                        error(errorNode, Diagnostics.Output_file_0_has_not_been_built_from_source_file_1, redirect, resolvedModule.resolvedFileName);
+                        return undefined;
                     }
                 }
 

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -2268,7 +2268,6 @@ namespace ts {
             if (refFile) {
                 const redirect = getProjectReferenceRedirect(fileName);
                 if (redirect) {
-                    ((refFile.redirectedReferences || (refFile.redirectedReferences = [])) as string[]).push(fileName);
                     fileName = redirect;
                     // Once we start redirecting to a file, we can potentially come back to it
                     // via a back-reference from another file in the .d.ts folder. If that happens we'll

--- a/src/compiler/types.ts
+++ b/src/compiler/types.ts
@@ -2740,12 +2740,6 @@ namespace ts {
         /* @internal */ resolvedModules?: Map<ResolvedModuleFull | undefined>;
         /* @internal */ resolvedTypeReferenceDirectiveNames: Map<ResolvedTypeReferenceDirective | undefined>;
         /* @internal */ imports: ReadonlyArray<StringLiteralLike>;
-        /**
-         * When a file's references are redirected due to project reference directives,
-         * the original names of the references are stored in this array
-         */
-        /* @internal*/
-        redirectedReferences?: ReadonlyArray<string>;
         // Identifier only if `declare global`
         /* @internal */ moduleAugmentations: ReadonlyArray<StringLiteral | Identifier>;
         /* @internal */ patternAmbientModules?: PatternAmbientModule[];
@@ -3074,6 +3068,7 @@ namespace ts {
         getSourceFiles(): ReadonlyArray<SourceFile>;
         getSourceFile(fileName: string): SourceFile | undefined;
         getResolvedTypeReferenceDirectives(): ReadonlyMap<ResolvedTypeReferenceDirective | undefined>;
+        getProjectReferenceRedirect(fileName: string): string | undefined;
 
         readonly redirectTargetsMap: RedirectTargetsMap;
     }

--- a/src/testRunner/unittests/config/projectReferences.ts
+++ b/src/testRunner/unittests/config/projectReferences.ts
@@ -284,6 +284,28 @@ namespace ts {
                 assertHasError("Issues a useful error", program.getSemanticDiagnostics(), Diagnostics.Output_file_0_has_not_been_built_from_source_file_1);
             });
         });
+
+        it("issues a nice error when the input file is missing when module reference is not relative", () => {
+            const spec: TestSpecification = {
+                "/alpha": {
+                    files: { "/alpha/a.ts": "export const m: number = 3;" },
+                    references: []
+                },
+                "/beta": {
+                    files: { "/beta/b.ts": "import { m } from '@alpha/a'" },
+                    references: ["../alpha"],
+                    options: {
+                        baseUrl: "./",
+                        paths: {
+                            "@alpha/*": ["/alpha/*"]
+                        }
+                    }
+                }
+            };
+            testProjectReferences(spec, "/beta/tsconfig.json", program => {
+                assertHasError("Issues a useful error", program.getSemanticDiagnostics(), Diagnostics.Output_file_0_has_not_been_built_from_source_file_1);
+            });
+        });
     });
 
     /**

